token-swap/program/fuzz/src/lib.rs

```diff
@@ -0,0 +1,586 @@
+            for account_info in account_infos.iter() {
+                if meta.pubkey == *account_info.key {
+                    let mut new_account_info = account_info.clone();
+                    for seeds in signers_seeds.iter() {
+                        let signer =
+                            Pubkey::create_program_address(&seeds, &spl_token_swap::id()).unwrap();
+                        if *account_info.key == signer {
+                            new_account_info.is_signer = true;
+                        }
+                    }
+                    new_account_infos.push(new_account_info);
+                }
+            }
+        }
+
+        spl_token::processor::Processor::process(
+            &instruction.program_id,
+            &new_account_infos,
+            &instruction.data,
+        )
+    }
+}
+
+fn test_syscall_stubs() {
+    use std::sync::Once;
+    static ONCE: Once = Once::new();
+
+    ONCE.call_once(|| {
+        program_stubs::set_syscall_stubs(Box::new(TestSyscallStubs {}));
+    });
+}
+
+fn do_process_instruction(instruction: Instruction, accounts: &[AccountInfo]) -> ProgramResult {
+    test_syscall_stubs();
+
+    // approximate the logic in the actual runtime which runs the instruction
+    // and only updates accounts if the instruction is successful
+    let mut account_data = accounts
+        .iter()
+        .map(AccountData::new_from_account_info)
+        .collect::<Vec<_>>();
+    let account_infos = account_data
+        .iter_mut()
+        .map(AccountData::into_account_info)
+        .collect::<Vec<_>>();
+    let res = if instruction.program_id == spl_token_swap::id() {
+        spl_token_swap::processor::Processor::process(
+            &instruction.program_id,
+            &account_infos,
+            &instruction.data,
+        )
+    } else {
+        spl_token::processor::Processor::process(
+            &instruction.program_id,
+            &account_infos,
+            &instruction.data,
+        )
+    };
+
+    if res.is_ok() {
+        let mut account_metas = instruction
+            .accounts
+            .iter()
+            .zip(accounts)
+            .map(|(account_meta, account)| (&account_meta.pubkey, account))
+            .collect::<Vec<_>>();
+        for account_info in account_infos.iter() {
+            for account_meta in account_metas.iter_mut() {
+                if account_info.key == account_meta.0 {
+                    let account = &mut account_meta.1;
+                    let mut lamports = account.lamports.borrow_mut();
+                    **lamports = **account_info.lamports.borrow();
+                    let mut data = account.data.borrow_mut();
+                    data.clone_from_slice(*account_info.data.borrow());
+                }
+            }
+        }
+    }
+    res
+}
+
+#[derive(Clone)]
+pub struct AccountData {
+    pub key: Pubkey,
+    pub lamports: u64,
+    pub data: Vec<u8>,
+    pub program_id: Pubkey,
+    pub is_signer: bool,
+}
```

Comment:
This is a bit hacky, but encompasses account data in an easy-to-use way.  Is there a better solution for this?